crate/operator/restore_backup.py

```diff
@@ -0,0 +1,925 @@
+    """
+    return "https" if "ssl" in spec["spec"]["cluster"] else "http"
+
+
+async def get_snapshot_tables(
+    cursor: Cursor, snapshot: str, logger: logging.Logger
+) -> List[Any]:
+    """
+    Returns a list of tables included in a snapshot.
+
+    :param cursor: A database cursor to a current and open database connection.
+    :param snapshot: The name of the snapshot where to lookup the tables.
+    :param logger: the logger on which we're logging
+    """
+    try:
+        logger.info("trying to get snapshot tables for snapshot %s", snapshot)
+        await cursor.execute(
+            "SELECT tables FROM sys.snapshots WHERE name=%s", (str(snapshot),)
+        )
+        row = await cursor.fetchone()
+        return row and row[0]
+    except ProgrammingError as e:
+        logger.warning("Failed to get snapshot tables", exc_info=e)
+        return []
+
+
+async def drop_repository(cursor: Cursor, repository: str, logger: logging.Logger):
+    """
+    Drops a backup repository if it exists.
+
+    :param cursor: A database cursor to a current and open database connection.
+    :param repository: The name of the repository to drop.
+    :param logger: the logger on which we're logging
+    """
+    try:
+        await cursor.execute(
+            "SELECT * FROM sys.repositories WHERE name=%s", (repository,)
+        )
+        row = await cursor.fetchone()
+        if row:
+            repository_ident = quote_ident(repository, cursor._impl)
+            await cursor.execute(f"DROP REPOSITORY {repository_ident}")
+    except ProgrammingError as e:
+        logger.warning("Failed to drop repository", exc_info=e)
+
+
+async def shards_recovery_in_progress(
+    cursor: Cursor,
+    logger: logging.Logger,
+    tables: List[str] = None,
+) -> Optional[List[int]]:
+    """
+    Returns a list of shards which have not fully recovered after an
+    operation of type ``SNAPSHOT``
+
+    :param cursor: A database cursor to a current and open database connection.
+    :param logger: the logger on which we're logging
+    :param tables: A list of tables which should be checked for shards in
+        progress.
+    """
+    tables = tables or []
+    in_progress = []
+    for t in tables:
+        (schema, table_name) = t.rsplit(".", 1)
+        logger.info(
+            "checking for shards in progress with schema %s and table name %s",
+            schema,
+            table_name,
+        )
+        await cursor.execute(
+            "SELECT id FROM sys.shards WHERE schema_name=%s "
+            "AND table_name=%s "
+            "AND recovery['type'] = 'SNAPSHOT' "
+            "AND recovery['size']['percent'] < 100 limit 1;",
+            (schema, table_name),
+        )
+        row = await cursor.fetchone()
+        if row:
+            logger.info("found shard with id %s ", row[0])
+            in_progress.append(int(row[0]))
+    return in_progress
+
+
+async def is_restore_complete(
```

Comment:
Doesn't the `ensure_no_restore_in_progress` ensures that a snapshot has completed already?

Reply:
true, I think this check is not necessary, I'll remove it..